src/instruction.rs

```diff
@@ -0,0 +1,268 @@
+//! Program instruction types.
+
+use solana_program::{
+    instruction::{AccountMeta, Instruction},
+    program_error::ProgramError,
+    pubkey::Pubkey,
+};
+
+/// Instructions supported by the Paladin Rewards program.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum PaladinRewardsInstruction {
+    /// Configures rewards for a mint that has been configured with the rewards
+    /// program as a transfer hook program.
+    ///
+    /// This instruction will:
+    ///
+    /// - Initialize a rewards account for the mint (distribution account),
+    ///   configured with the distribution addresses:
+    ///   - Active rewards
+    ///   - Piggy bank
+    ///   - Staked PAL rewards
+    /// - Initialize an active rewards account for the mint.
+    /// - Initialize the required accounts for the transfer hook.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    /// 0. `[w]` Distribution account.
+    /// 1. `[w]` Active rewards account.
+    /// 2. `[w]` Transfer hook extra account metas account.
+    /// 3. `[ ]` Token mint.
+    /// 4. `[s]` Mint authority.
+    InitializeMintRewardInfo {
+        piggy_bank_address: Pubkey,
+        staked_rewards_address: Pubkey,
+    },
+    /// Moves the active rewards to the distribution account and updates total
+    /// rewards.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    /// 0. `[w]` Active rewards account.
+    /// 1. `[w]` Distribution account.
+    /// 2. `[ ]` Token mint.
+    UpdateMintRewards,
+    /// Moves SOL rewards to the following parties:
+    ///
+    /// - 1%  Piggy bank
+    /// - 4%  Staked PAL (validators)
+    /// - 5%  All PAL holders
+    /// - 90% Leader who produces the block
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    /// 0. `[w]` Active rewards account.
+    /// 1. `[w]` Distribution account.
+    /// 2. `[w]` Piggy bank account.
+    /// 3. `[w]` Staked PAL rewards account.
+    /// 4. `[w]` Leader account.
+    DistributeRewards,
+    /// Initializes holder reward info by storing the last seen total rewards
+    /// in the distribution account.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    /// 0. `[w]` Holder rewards account.
+    /// 1. `[ ]` PAL token account.
+    /// 2. `[ ]` PAL token mint.
+    /// 3. `[s]` PAL token account owner.
```

Comment:
I don't think this should take a signature from the token account owner -- it makes it impossible to create accounts for others